### DISCUSSION TITLE
维护一个koishi bot项目的最佳实践是什么样的？

### DISCUSSION THREAD

**Discussion by axiangcoding** [koishijs/koishi#1269] [Q&A] [answered]
我是koishi的新手，所以我对koishi项目的最佳实践不太了解。

根据 https://koishi.chat/zh-CN/guide/develop/workspace.html 的描述，各种业务均是通过插件来实现。

那么假设我要开发一个拥有若干业务的机器人，我按照文档创建了一个koishi的模板项目，那么我的git根目录是我的koishi模板项目的根目录，还是插件文件夹的根目录？

**Comment by MaikoTan** [comment]
I think the [document here](https://koishi.chat/zh-CN/cookbook/design/workspace.html) would solve several your problems, in general we recommend to make your workspace root as the git root, and you can put other workspaces or plugins as [git submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) in the `external/` folder.

我认为[这里的文档](https://koishi.chat/zh-CN/cookbook/design/workspace.html)可以解决您的几个问题，一般来说我们建议将您的工作空间根目录设置为git根目录，并且您可以将其他工作区或插件作为 [git submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) 放在 `external/` 文件夹中。

**Reply by axiangcoding** [reply]
thanks for the reply.

**Comment by ilharp** [comment]
对工作区启用版本控制可以使用版本控制托管你的机器人的配置，包括插件选项、本地化文件和其他配置。

对插件启用版本控制可以让其他人快速找到和开发单个插件。

新建的 Koishi 项目默认对所有插件启用版本控制。在此基础上，你可以对工作区或任意插件启用或停用版本控制，一切都可按照自己的喜好来。

**Reply by axiangcoding** [reply]
如果我期望视整个机器人（包括了若干自定义插件或者公开插件）为整体，并以整个机器人为单位部署机器人，那是否对工作区启用版本控制即可？

**Reply by ilharp** [reply]
是的。这样做的问题在于，其他人会更难搜索到你开发和发布的插件（因为直接拿插件名称上 GitHub 搜索 repo 搜索不到）。

如果没有这样的需求或不想考虑这样的问题，那么可以直接对工作区启用版本控制。整个机器人项目会作为一个 repo 被 Git 托管和呈现。

**Reply by axiangcoding** [reply]
我明白了。那我如果期望插件也做为一个整体发布，那我的最佳实践是否是这样：
- 自定义插件单独仓库维护，发布npm
- 整体机器人单独仓库，使用npm引入插件的方式使用插件

这样是否就达成了插件可发布，仓库可维护的想法

**Reply by ilharp** [reply, marked as answer]
可以的。甚至，你还可以使用 Git 提供的 Submodule 或 Subtree 功能，将插件仓库作为机器人仓库的子仓库，这样两个仓库仍然具有独立的版本控制，但形式上插件是机器人项目的子文件夹，可以在同一个工作区里同时开发机器人配置和插件。

如果你尚不熟悉 Submodule 或 Subtree（二者坑都 **非常** 多），你可以先按照你的思路，分别使用两个仓库管理，然后使用 npm 安装插件。

**Comment by shigma** [comment]
正如 @MaikoTan 的回答，[深入工作区](https://koishi.chat/zh-CN/cookbook/design/workspace.html) 可以部分解决你的问题。然而设计原理部分目前还在活跃更新中，「整合包开发」一节还没写出来，或许你可以到时候再看看 🤣

**Reply by axiangcoding** [reply]
“整合包开发”大概可以什么时候提供文档呢

**Reply by shigma** [reply]
https://koishi.chat/zh-CN/cookbook/design/workspace.html

https://koishi.chat/zh-CN/cookbook/design/bundle.html
